stan/math/prim/functor/closure_adapter.hpp

```diff
@@ -0,0 +1,231 @@
+#ifndef STAN_MATH_PRIM_FUNCTOR_CLOSURE_ADAPTER_HPP
+#define STAN_MATH_PRIM_FUNCTOR_CLOSURE_ADAPTER_HPP
+
+#include <stan/math/prim/meta/is_stan_closure.hpp>
+#include <stan/math/prim/meta/return_type.hpp>
+#include <ostream>
+
+namespace stan {
+namespace math {
+
+template <typename F>
+struct empty_closure {
+  using captured_scalar_t__ = double;
+  using ValueOf__ = empty_closure<F>;
+  using CopyOf__ = empty_closure<F>;
+  F f_;
+
+  explicit empty_closure(const F& f) : f_(f) {}
+
+  template <typename... Args>
+  auto operator()(std::ostream* msgs, Args... args) const {
+    return f_(args..., msgs);
+  }
+  size_t count_vars__() const { return 0; }
+  auto value_of__() const { return ValueOf__(f_); }
+  auto copy_of__() const { return CopyOf__(f_); }
+  auto deep_copy_vars__() const { return CopyOf__(f_); }
+  void zero_adjoints__() const {}
+  double* accumulate_adjoints__(double* dest) const { return dest; }
+  template <typename Vari>
+  Vari** save_varis(Vari** dest) const {
+    return dest;
+  }
+};
+
+template <bool Ref, typename F, typename T>
+struct one_arg_closure {
+  using captured_scalar_t__ = return_type_t<T>;
+  using ValueOf__
+      = one_arg_closure<false, F, decltype(value_of(std::declval<T>()))>;
+  using CopyOf__ = one_arg_closure<false, F, T>;
+  F f_;
+  capture_type_t<T, Ref> s_;
+
+  explicit one_arg_closure(const F& f, const T& s) : f_(f), s_(s) {}
+
+  template <typename... Args>
+  auto operator()(std::ostream* msgs, Args... args) const {
+    return f_(s_, args..., msgs);
```

Comment:
Having functions where the variadic arguments aren't last makes writing the C++ painful (cause no type deduction). Could it go `captured_args..., msgs, args...`?